IntegrationTests/tests_04_performance/test_01_resources/template/Sources/SwiftBootstrap/SwiftMain.swift

```diff
@@ -50,6 +50,85 @@ private final class SimpleHTTPServer: ChannelInboundHandler {
     }
 }
 
+private struct PingPongFailure: Error, CustomStringConvertible {
+    public var description: String
+
+    init(problem: String) {
+        self.description = problem
+    }
+}
+
+private final class PingHandler: ChannelInboundHandler {
+    typealias InboundIn = ByteBuffer
+    typealias OutboundOut = ByteBuffer
+
+    private var pingBuffer: ByteBuffer!
+    private let numberOfRequests: Int
+    private var remainingNumberOfRequests: Int
+    private let allDone: EventLoopPromise<Void>
+
+    public init(numberOfRequests: Int, eventLoop: EventLoop) {
+        self.numberOfRequests = numberOfRequests
+        self.remainingNumberOfRequests = numberOfRequests
+        self.allDone = eventLoop.newPromise()
+    }
+
+    public func channelActive(ctx: ChannelHandlerContext) {
+        self.pingBuffer = ctx.channel.allocator.buffer(capacity: 4)
+        self.pingBuffer.write(staticString: "ping")
+
+        ctx.writeAndFlush(self.wrapOutboundOut(self.pingBuffer), promise: nil)
+    }
+
+    public func channelRead(ctx: ChannelHandlerContext, data: NIOAny) {
+        var buf = self.unwrapInboundIn(data)
+        if buf.readableBytes == 4 &&
```

Comment:
This is not correct... as tcp is Stream based and can be fragmented we need to use a ByteToMessageDecoder here and keep
Buffering if the readable bytes are smaller then 4

Reply:
Or just keep state by ourself

Reply:
Oh sure but I mean it’s loopback and only ever 4 bytes in transit. I thought about just making it 1 byte to not have this problem but then thought that’s over the top as that’ll actually not happen.
But happy to change it to 1 byte.
Buffering won’t work as we count every allocation and if we were to buffer we’d add allocations.
I think 4 bytes as is is fine but can change it to one byte if you want.

Reply:
I think 1 byte would be a safer bet here.

Reply:
fair enough. Didn't doo it because I thought sending `ping` and `pong` around is much more fun than `x` and `y`. Can't even send emoji around now 😢